core/src/main/java/io/confluent/kafka/schemaregistry/rest/resources/SubjectVersionsResource.java

```diff
@@ -624,18 +629,46 @@ public void modifyTags(
     }
     schemaRegistry.getCompositeUpdateRequestHandler().handle(schema, request);
 
-    if (request.getRuleSet() != null) {
-      try {
-        request.getRuleSet().validate();
-      } catch (RuleException e) {
-        throw new RestInvalidRuleSetException(e.getMessage());
-      }
-    }
-
     Map<String, String> headerProperties = requestHeaderBuilder.buildRequestHeaders(
         headers, schemaRegistry.config().whitelistHeaders());
     RegisterSchemaResponse registerSchemaResponse;
     try {
+      RuleSet ruleSet = request.getRuleSet();
+      if (request.getRulesToMerge() != null || request.getRulesToRemove() != null) {
+        int oldVersion = request.getNewVersion() != null ? request.getNewVersion() - 1 : -1;
+        Schema oldSchema = schemaRegistry.get(subject, oldVersion, false);
+        // Use the previous ruleSet instead of the passed in one
+        ruleSet = oldSchema != null ? oldSchema.getRuleSet() : null;
+        if (request.getRulesToMerge() != null) {
+          ruleSet = mergeRuleSets(ruleSet, request.getRulesToMerge());
```

Comment:
Does this change the ordering of the rules? If so does that matter? I remember you mentioning that the order of the ruse matter. 

It seems like merging the rules would remove any existing role that has the same name as any rule in `rulesToMerge`, then add `rulesToMerge` at the end of the rules list.